address remaining issues (#118)
Atcold/pytorch-Deep-Learning: docs/en/week06/06-3.md

<center>
<img src="{{site.baseurl}}/images/week06/06-3/memory_cell_vis.png" style="zoom: 40%; background-color:#DCDCDC;"/><br>
<b>Figure 19:</b>Visualization of the Memory Cell
</center>

<center>
<img src="{{site.baseurl}}/images/week06/06-3/lstm_4.png" style="zoom: 40%; background-color:#DCDCDC;"/><br>
<b>Figure 20:</b>LSTM Architecture - Reset Memory
</center>

<center>
<img src="{{site.baseurl}}/images/week06/06-3/lstm_keep_memory.png" style="zoom: 40%; background-color:#DCDCDC;"/><br>
<b>Figure 21:</b>LSTM Architecture - Keep Memory
</center>

<center>
<img src="{{site.baseurl}}/images/week06/06-3/lstm_write_memory.png" style="zoom: 40%; background-color:#DCDCDC;"/><br>
<b>Figure 22:</b>LSTM Architecture - Write Memory
</center>

## Notebook Examples

### Sequence Classification

> The goal is to classify sequences. Elements and targets are represented locally (input vectors with only one non-zero bit). The sequence starts with an $B$, ends with a $E$ (the “trigger symbol”), and otherwise consists of randomly chosen symbols from the set $\{a, b, c, d\}$ except for two elements at positions $t1$ and $t2$ that are either $X$ or $Y$. For the DifficultyLevel.HARD case, the sequence length is randomly chosen between 100 and 110, $t1$ is randomly chosen between 10 and 20, and $t2$ is randomly chosen between 50 and 60. There are 4 sequence classes $Q, R, S$, and $U$, which depend on the temporal order of $X$ and $Y$. The rules are: `X, X -> Q`; `X, Y -> R`; `Y, X -> S`; `Y, Y -> U`.

Reviewer: ```suggestion
The goal is to classify sequences. Elements and targets are represented locally (input vectors with only one non-zero bit). The sequence **b**egins with an `B`, **e**nds with a `E` (the “trigger symbol”), and otherwise consists of randomly chosen symbols from the set `{a, b, c, d}` except for two elements at positions $t_1$ and $t_2$ that are either `X` or `Y`. For the `DifficultyLevel.HARD` case, the sequence length is randomly chosen between 100 and 110, $t_1$ is randomly chosen between 10 and 20, and $t_2$ is randomly chosen between 50 and 60. There are 4 sequence classes `Q`, `R`, `S`, and `U`, which depend on the temporal order of `X` and `Y`. The rules are: `X, X -> Q`; `X, Y -> R`; `Y, X -> S`; `Y, Y -> U`.
```

1). Dataset Exploration

> The return type from a data generator is a tuple with length 2. The first item in the tuple is the batch of sequences with shape $(32, 9, 8)$. This is the data going to be fed into the network. There are eight different symbols in each row $(X, Y, a,b, c, d, B, E)$. Each row is a one-hot vector. Sequence of rows represents sequence of symbols. The first all-zero row is padding. We use padding when the length of sequence is shorter than the maximum length in the batch.  The second item in the tuple is the corresponding batch of class labels with shape $(32, 4)$, since we have 4 classes $(Q, R, S, and U)$. The first sequence is: `BbXcXcbE`. Then its decoded class label is $[1, 0, 0, 0]$, corresponding to $Q$.

Reviewer: ```suggestion
The return type from a data generator is a tuple with length 2. The first item in the tuple is the batch of sequences with shape $(32, 9, 8)$. This is the data going to be fed into the network. There are eight different symbols in each row (`X`, `Y`, `a`, `b`, `c`, `d`, `B`, `E`). Each row is a one-hot vector. Sequence of rows represents sequence of symbols. The first all-zero row is padding. We use padding when the length of sequence is shorter than the maximum length in the batch.  The second item in the tuple is the corresponding batch of class labels with shape $(32, 4)$, since we have 4 classes (`Q`, `R`, `S`, and `U`). The first sequence is: `BbXcXcbE`. Then its decoded class label is $[1, 0, 0, 0]$, corresponding to `Q`.
```

* Perform the forward pass of the model
* Compute the loss
* Zero the gradient cache
* Backpropagate to compute the partial derivative of loss with regard to parameters
* Step in the opposite direction of gradient

<center>
<img src="{{site.baseurl}}/images/week06/06-3/train_test_easy.png" style="zoom: 40%; background-color:#DCDCDC;"/><br>
<b>Figure 24:</b>Simple RNN vs LSTM - 10 Epochs
</center>

With easy level of difficulty, RNN gets 50% accuracy while LSTM gets 100% after 10 epoches. But LSTM has four times more weights than RNN and has two hidden layers, so it is not a fair comparison. After 100 epoches, RNN also gets 100% accuracy, but it takes longer than LSTM.

<center>
<img src="{{site.baseurl}}/images/week06/06-3/train_test_hard.png" style="zoom: 40%; background-color:#DCDCDC;"/><br>
<b>Figure 25:</b>Simple RNN vs LSTM - 100 Epochs
</center>

If we increase the difficulty of the training part (using longer sequences), we will see the RNN fails while LSTM keeps working.

<center>
<img src="{{site.baseurl}}/images/week06/06-3/hidden_state_lstm.png" style="zoom: 40%; background-color:#DCDCDC;"/><br>
<b>Figure 26:</b>Visualization of Hidden State Value
</center>

> The above visualization is drawing the value of hidden state over time in LSTM. We will send the inputs through a hyperbolic tangent, such that if the input is below -2.5, it will be mapped to -1, and if it is above 2.5, it will be mapped to 1. So in this case, we can see the specific hidden layer picked on $X$ (fifth row in the picture) and then it became red until we got the other $X$. So, the fifth hidden unit of the cell is triggered by observing the $X$ and goes quiet after seeing the other $X$. This allows us to recognize the class of sequence.

Reviewer: ```suggestion
The above visualization is drawing the value of hidden state over time in LSTM. We will send the inputs through a hyperbolic tangent, such that if the input is below $-2.5$, it will be mapped to $-1$, and if it is above $2.5$, it will be mapped to $1$. So in this case, we can see the specific hidden layer picked on `X` (fifth row in the picture) and then it became red until we got the other `X`. So, the fifth hidden unit of the cell is triggered by observing the `X` and goes quiet after seeing the other `X`. This allows us to recognize the class of sequence.
```
